### PR TITLE
New Photometric Stereo software

### PR DESCRIPTION
**Pull request by jmelou**

### CODE REVIEW

File: src/aliceVision/photometricStereo/photometricDataIO.cpp [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+#include <sstream>
+#include <fstream>
+
+#include "photometricDataIO.hpp"
+
+void loadLightIntensities(const std::string& intFileName, const std::vector<int>& usedPictures, std::vector<std::array<float, 3>>& intList)
+{
+    std::stringstream stream;
+    std::string line;
+    float x,y,z;
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.cpp:13]:
```suggestion

```

File: src/aliceVision/photometricStereo/photometricDataIO.cpp [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+		
+        while(!intFile.eof())
+        {
+			std::getline(intFile,line);
+			stream.clear();
+			stream.str(line);
+				
+			if(std::find(usedPictures.begin(), usedPictures.end(), cpt) != usedPictures.end())
+			{
+				stream >> x >> y >> z;
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.cpp:30]:
```suggestion
				float x,y,z;
				stream >> x >> y >> z;
```

File: src/aliceVision/photometricStereo/photometricDataIO.cpp [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+void loadLightIntensities(const std::string& intFileName, const std::vector<int>& usedPictures, std::vector<std::array<float, 3>>& intList)
+{
+    std::stringstream stream;
+    std::string line;
+    float x,y,z;
+    
+    std::fstream intFile;
+    intFile.open(intFileName, std::ios::in);
+    
+    if (intFile.is_open())
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.cpp:18]: prefer early exit conditions
```suggestion
    if (!intFile.is_open())
    {
        ALICEVIOSION_LOG_ERROR(....);
        throw ....;
    }
```

File: src/aliceVision/photometricStereo/photometricDataIO.cpp [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    {
+		int cpt = 1;
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.cpp:20]: use spaces to indent

File: src/aliceVision/photometricStereo/photometricDataIO.cpp [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+void loadLightDirections(const std::string& dirFileName, const std::vector<int>& usedPictures, Eigen::MatrixXf& lightMat)
+{
+    std::stringstream stream;
+    std::string line;
+    float x,y,z;
+    
+    std::fstream dirFile;
+    dirFile.open(dirFileName, std::ios::in);
+    
+    if (dirFile.is_open())
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.cpp:50]: same here

File: src/aliceVision/photometricStereo/photometricDataIO.hpp [hunk 6]
```diff
@@ -0,0 +1,24 @@
+#ifndef DATAIO_HPP_INCLUDED
+#define DATAIO_HPP_INCLUDED
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.hpp:2]:
```suggestion
#pragma once
```

File: src/aliceVision/photometricStereo/photometricDataIO.hpp [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+void image2PsMatrix(const aliceVision::image::Image<aliceVision::image::RGBfColor>& imageIn, Eigen::MatrixXf& imageOut);
+
+void applyMask(const Eigen::MatrixXf& inputMatrix, const std::vector<int>& maskIndexes, Eigen::MatrixXf& maskedMatrix);
+
+void normals2picture(const Eigen::MatrixXf& normalsMatrix, aliceVision::image::Image<aliceVision::image::RGBfColor>& normalsIm);
+
+void convertNormalMap2png(const aliceVision::image::Image<aliceVision::image::RGBfColor>& normalsIm, aliceVision::image::Image<aliceVision::image::RGBColor>& normalsImPNG);
+
+#endif // DATAIO_HPP_INCLUDED
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.hpp:24]:
```suggestion

```

File: src/aliceVision/photometricStereo/photometricDataIO.hpp [hunk 8]
```diff
@@ -0,0 +1,24 @@
+#ifndef DATAIO_HPP_INCLUDED
+#define DATAIO_HPP_INCLUDED
+#include <string>
+#include <vector>
+
+void loadLightIntensities(const std::string& intFileName, const std::vector<int>& usedPictures, std::vector<std::array<float, 3>>& intList);
```
Comment (simogasp) [inline src/aliceVision/photometricStereo/photometricDataIO.hpp:6]: add doxygen documentation

**Review by simogasp — CHANGES_REQUESTED** [review]
please make sure you are using 4 spaces to indent the code and not tab

File: src/aliceVision/lightingEstimation/lightingCalibration.cpp [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+    {
+        const fs::path imagePath = fs::path(viewId.getImagePath());
+
+        if (!boost::algorithm::icontains(imagePath.stem().string(), "ambiant"))
+        {
+            bpt::ptree lightTree;
+            if (saveAsModel)
+            {
+                lightTree.put("lightId", imgCpt);
+                lightTree.put("type", "directionnal");
```
Comment (fabiencastan) [inline src/aliceVision/lightingEstimation/lightingCalibration.cpp:283]:
```suggestion
                lightTree.put("type", "directional");
```

File: src/aliceVision/lightingEstimation/lightingCalibration.cpp [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+            }
+            else
+            {
+                IndexT index = viewId.getViewId();
+
+                lightTree.put("viewId", index);
+                lightTree.put("type", "directionnal");
```
Comment (fabiencastan) [inline src/aliceVision/lightingEstimation/lightingCalibration.cpp:290]:
```suggestion
                lightTree.put("type", "directional");
```

File: src/aliceVision/lightingEstimation/lightingCalibration.cpp [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+    std::vector<std::string> imageList;
+    std::string pictureFolder = inputPath + "PS_Pictures/";
+    // getPicturesNames(pictureFolder, imageList);  ???
+
+    // std::array<float, 3> sphereParam;
+    // image::Image<float> sphereMask;
+
+    // detectSphere(imageList, sphereParam, sphereMask);
+    // lightCalibration(imageList, sphereParam, outputPath);
+}
```
Comment (fabiencastan) [inline src/aliceVision/lightingEstimation/lightingCalibration.cpp:48]: ??

Comment (fabiencastan) [inline src/aliceVision/lightingEstimation/lightingCalibration.cpp:48]:
```suggestion
```

File: src/aliceVision/sphereDetection/sphereDetection.hpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+// Boost
+#include <boost/filesystem.hpp>
+
+// namespaces
+namespace fs = boost::filesystem;
+
+namespace aliceVision {
+namespace sphereDetection {
+
+struct prediction
```
Comment (fabiencastan) [inline src/aliceVision/sphereDetection/sphereDetection.hpp:28]: CamelCase

File: src/software/pipeline/main_lightingCalibration.cpp [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    cmdline.add(optionalParams);
+
+    if (!cmdline.execute(argc, argv))
+    {
+        return EXIT_FAILURE;
+    }
+
+    if (fs::is_directory(inputPath))
+    {
+        std::cout << "Directory input : WIP" << std::endl;
```
Comment (fabiencastan) [inline src/software/pipeline/main_lightingCalibration.cpp:75]: Throw with an explicit error message.

File: src/software/pipeline/main_photometricStereo.cpp [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+    if (!cmdline.execute(argc, argv))
+    {
+        return EXIT_FAILURE;
+    }
+
+    // If the path to light data is empty, set it to inputPath :
+    if (pathToLightData.compare("") && fs::is_directory(inputPath))
+    {
+        std::cout << "Warning : path to light data has been set to inputpath folder" << std::endl;
```
Comment (fabiencastan) [inline src/software/pipeline/main_photometricStereo.cpp:92]: ALICEVISION_LOG_WARNING

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**